topics/data-science/tutorials/python-loops/tutorial.md

```diff
@@ -0,0 +1,676 @@
+
+## Things You Can Loop Over
+
+You can loop over characters in a string
+
+```python
+dna_string = 'ACTGGTCATCG'
+for base in dna_string:
+    print(base)
+```
+
+You can loop over lists:
+
+```python
+cast = ['Elphaba', 'Glinda', 'Fiyero', 'Nessarose']
+for character in cast:
+    print(character)
+```
+
+## Indentation
+
+The first line of the `for` loop must end with a colon, and the body must be indented with *four spaces*. Many editors do this automatically for you and even convert <kbd>Tab</kbd>s into 4 spaces.
+
+> ### {% icon tip %} Tip: Blocks in Python
+>
+> The colon at the end of the first line signals the start of a *block* of statements.
+>
+> ```
+> for x in y:
+>     print(x)
+> ```
+>
+> or
+>
+> ```
+> if x > 10:
+>     print(x)
+> ```
+>
+> or even further nesting is possible:
+>
+> ```
+> for x in y:
+>     if x > 10:
+>         print(x)
+> ```
+>
+{: .tip}
+
+The indentation is in fact, quite necessary. Notice how this fails:
+
+```python
+for number in [2, 3, 5]:
+print(number)
+```
+
+And, likewise, this:
+
+```python
+patient1 = "z2910"
+  patient2 = "y9583"
+```
+
+## Variable Naming
+
+Loop variables can be called anything, `i`, `j`, and `k` are very commong defaults due to their long history of use in other programing languages.
+As with all variables, loop variables are: Created on demand, and Meaningless; their names can be anything at all.
+
+```python
+for kitten in [2, 3, 5]:
+    print(kitten)
+```
+
+But *meaningless* is bad for variable names, and whenever possible, we should strive to pick useful, accurate variable names that help use remember what's going on:
+
+```
+for sequence in sequences:
+    print()
+for patient in clinic_patients:
+    print()
+for nucleotide in dna_sequence:
+    print()
+```
+
+## Range
+
+You can use `range` to iterate over a sequence of numbers. This is a built in function (check `help(range)`!) so it's always available even if you don't `import` anything. The range produced is non-inclusive: `range(N)` is the numbers `0` to `N-1`, so the result will be exactly the length you requested.
+
+```python
+for number in range(10):
+    print(number)
+```
+
+> ### {% icon tip %} Tip: Iterables can be weird
+> In python `range` is a special type of iterable: none of the numbers are created until we need them.
+>
+> ```python
+> print(range(5))
+> print(range(-3, 8)[0:4])
+> ```
+>
+> The easiest way to see what numbers are actually in there is to convert it to a `list`:
+>
+> ```python
+> print(list(range(5)))
+> print(list(range(-3, 8)))
+> print(list(range(0, 10, 2)))
+> ```
+{: .tip}
+
+## Accumulation
+
+In programming you'll often want to accumulate some values: counting things (or "accumulating"). The pattern consists of creating a variable to store your result, running a loop over some data, and in that loop, adding to the variable for your result.
+
+```python
+# Sum the first 10 integers.
+total = 0
+for number in range(1, 11):
+    total = total + (number)
```

Comment:
```suggestion
    total = total + (number)
    print(f"in loop total: {{ total}} ")
```

Reply:
See the lines just below this (but yes, I agree :) ) my intention was just that they type it themselves a bit.

```
But how did we get that result? We can add some "debugging" lines to the above code to figure out how we got to that result. Try adding the following line in the above loop

    print(f'Currently {number}, our total is {total}')
```

Reply:
ahha. nvm then :)